motion_primitives_forward_controller/src/motion_primitives_forward_controller.cpp

```diff
@@ -0,0 +1,554 @@
+    }
+    else if (arg.argument_name == "acceleration")
+    {
+      (void)command_interfaces_[23].set_value(arg.argument_value);
+    }
+    else if (arg.argument_name == "move_time")
+    {
+      (void)command_interfaces_[24].set_value(arg.argument_value);
+    }
+    else
+    {
+      RCLCPP_WARN(
+        get_node()->get_logger(), "Unknown additional argument: %s", arg.argument_name.c_str());
+    }
+  }
+  return true;
+}
+
+rclcpp_action::GoalResponse MotionPrimitivesForwardController::goal_received_callback(
+  const rclcpp_action::GoalUUID &, std::shared_ptr<const ExecuteMotion::Goal> goal)
+{
+  RCLCPP_INFO(get_node()->get_logger(), "Received goal request");
+
+  const auto & primitives = goal->trajectory.motions;
+
+  if (primitives.empty())
+  {
+    RCLCPP_WARN(get_node()->get_logger(), "Goal rejected: no motion primitives provided.");
+    return rclcpp_action::GoalResponse::REJECT;
+  }
+
+  for (size_t i = 0; i < primitives.size(); ++i)
+  {
+    const auto & primitive = primitives[i];
+
+    switch (primitive.type)
+    {
+      case MotionType::LINEAR_JOINT:
+        RCLCPP_INFO(get_node()->get_logger(), "Primitive %zu: LINEAR_JOINT (PTP)", i);
+        if (primitive.joint_positions.empty())
+        {
+          RCLCPP_ERROR(
+            get_node()->get_logger(),
+            "Primitive %zu invalid: LINEAR_JOINT requires joint_positions.", i);
+          return rclcpp_action::GoalResponse::REJECT;
+        }
+        break;
+
+      case MotionType::LINEAR_CARTESIAN:
+        RCLCPP_INFO(get_node()->get_logger(), "Primitive %zu: LINEAR_CARTESIAN (LIN)", i);
+        if (primitive.poses.empty())
+        {
+          RCLCPP_ERROR(
+            get_node()->get_logger(),
+            "Primitive %zu invalid: LINEAR_CARTESIAN requires at least one pose.", i);
+          return rclcpp_action::GoalResponse::REJECT;
+        }
+        break;
+
+      case MotionType::CIRCULAR_CARTESIAN:
+        RCLCPP_INFO(get_node()->get_logger(), "Primitive %zu: CIRCULAR_CARTESIAN (CIRC)", i);
+        if (primitive.poses.size() != 2)
+        {
+          RCLCPP_ERROR(
+            get_node()->get_logger(),
+            "Primitive %zu invalid: CIRCULAR_CARTESIAN requires exactly two poses.", i);
+          return rclcpp_action::GoalResponse::REJECT;
+        }
+        break;
+
+      default:
+        RCLCPP_ERROR(
+          get_node()->get_logger(), "Primitive %zu invalid: unknown motion type %u.", i,
+          primitive.type);
+        return rclcpp_action::GoalResponse::REJECT;
+    }
+  }
+
+  return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
+}
+
+rclcpp_action::CancelResponse MotionPrimitivesForwardController::goal_cancelled_callback(
+  const std::shared_ptr<rclcpp_action::ServerGoalHandle<ExecuteMotion>>)
+{
+  RCLCPP_INFO(get_node()->get_logger(), "Canceling goal");
+
+  std::lock_guard<std::mutex> guard(command_mutex_);
+  reset_command_interfaces();
+  robot_stopped_ = true;
```

Comment:
nitpicking: Isn't that variable encoding "robot stop requested" rather than "robot stopped"?